File: clients/rust/src/limits.rs
//! Protocol limits. Callers may tighten the defaults, never widen the maxima.

use serde::{Deserialize, Serialize};

/// Per-session capacity ceilings, named as they appear on the wire.
/// `limits` is the one object on the wire that grows between protocol
/// versions, so unknown fields are ignored rather than rejected: a client that
/// refused a ceiling it had never heard of would drop the channel every time
/// the protocol gained one.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Limits {
    /// Per-frame byte ceiling, header excluded.
    pub max_frame_bytes: usize,
    /// Serialised size ceiling for one snapshot.
    pub max_snapshot_bytes: usize,
    /// Node and root-id count ceiling per snapshot.
    pub max_nodes: usize,
    /// Structural nesting ceiling, roots at depth 1.
    pub max_depth: usize,
    /// UTF-8 byte ceiling for any single string.
    pub max_string_bytes: usize,
    /// Ceiling on `labelledBy`/`describedBy`/`textRanges` entries.
    pub max_relation_targets: usize,
    /// Frames the driver buffers before applying back-pressure.
    pub max_queued_frames: usize,
    /// Concurrent waiters the driver will track.
    pub max_pending_waiters: usize,
    /// Concurrent sessions the driver will hold open.
    pub max_sessions: usize,
    /// Byte ceiling for one serialised application log record.
    pub max_log_record_bytes: usize,
    /// Log records the driver buffers per session before evicting the oldest.
    pub max_log_queue: usize,
}

/// What an adapter assumes until `hello-ack` says otherwise.
pub const DEFAULT_LIMITS: Limits = Limits {
    max_frame_bytes: 1024 * 1024,
    max_snapshot_bytes: 2 * 1024 * 1024,
    max_nodes: 5_000,
    max_depth: 64,
    max_string_bytes: 16 * 1024,
    max_relation_targets: 64,
    max_queued_frames: 32,
    max_pending_waiters: 256,
    max_sessions: 16,
    max_log_record_bytes: 32 * 1024,
    max_log_queue: 1_000,
};

/// The widest configuration either side may accept.
pub const ABSOLUTE_LIMITS: Limits = Limits {
    max_frame_bytes: 8 * 1024 * 1024,
    max_snapshot_bytes: 8 * 1024 * 1024,
    max_nodes: 50_000,
    max_depth: 256,
    max_string_bytes: 256 * 1024,
    max_relation_targets: 1_024,
    max_queued_frames: 256,
    max_pending_waiters: 4_096,
    max_sessions: 128,
    max_log_record_bytes: 256 * 1024,
    max_log_queue: 10_000,
};

/// Milliseconds a driver waits for a `hello` before settling the session as
/// generic (non-semantic).
pub const DEFAULT_NEGOTIATION_MS: u64 = 250;

impl Default for Limits {
    fn default() -> Self {
        DEFAULT_LIMITS
    }
}

File: clients/go/protocol/limits.go
package protocol

// Limits are the per-session capacity ceilings. Callers may tighten the
// defaults; they can never widen the absolute maxima.
//
// Field names are the wire names, so a Limits value marshals straight into a
// hello-ack payload.
type Limits struct {
	MaxFrameBytes      int `json:"maxFrameBytes"`
	MaxSnapshotBytes   int `json:"maxSnapshotBytes"`
	MaxNodes           int `json:"maxNodes"`
	MaxDepth           int `json:"maxDepth"`
	MaxStringBytes     int `json:"maxStringBytes"`
	MaxRelationTargets int `json:"maxRelationTargets"`
	MaxQueuedFrames    int `json:"maxQueuedFrames"`
	MaxPendingWaiters  int `json:"maxPendingWaiters"`
	MaxSessions        int `json:"maxSessions"`
	// MaxLogRecordBytes bounds one serialised application log record.
	MaxLogRecordBytes int `json:"maxLogRecordBytes"`
	// MaxLogQueue is how many log records the driver buffers per session.
	MaxLogQueue int `json:"maxLogQueue"`
}

// DefaultLimits is what an adapter assumes until hello-ack says otherwise.
var DefaultLimits = Limits{
	MaxFrameBytes:      1 * 1024 * 1024,
	MaxSnapshotBytes:   2 * 1024 * 1024,
	MaxNodes:           5000,
	MaxDepth:           64,
	MaxStringBytes:     16 * 1024,
	MaxRelationTargets: 64,
	MaxQueuedFrames:    32,
	MaxPendingWaiters:  256,
	MaxSessions:        16,
	MaxLogRecordBytes:  32 * 1024,
	MaxLogQueue:        1000,
}

// AbsoluteLimits is the widest configuration any side may accept.
var AbsoluteLimits = Limits{
	MaxFrameBytes:      8 * 1024 * 1024,
	MaxSnapshotBytes:   8 * 1024 * 1024,
	MaxNodes:           50000,
	MaxDepth:           256,
	MaxStringBytes:     256 * 1024,
	MaxRelationTargets: 1024,
	MaxQueuedFrames:    256,
	MaxPendingWaiters:  4096,
	MaxSessions:        128,
	MaxLogRecordBytes:  256 * 1024,
	MaxLogQueue:        10000,
}

// DefaultNegotiationMillis is the window a driver waits for a hello before it
// settles the session as generic (non-semantic).
const DefaultNegotiationMillis = 250
